tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/AbstractOperation.java

```diff
@@ -84,8 +84,9 @@ public String toString() {
    *
    * <p>This is only supported in an eager execution environment.
    *
+   * @param scope the {@link TensorScope} to create the tensor in
    * @param outputIdx index of the output of this operation
    * @return output tensor
    */
-  abstract Tensor tensor(int outputIdx);
+  abstract Tensor tensor(TensorScope scope, int outputIdx);
```

Comment:
I think the scope is going to end up by convention as the last argument. That allows it to be made optional more naturally.

Reply:
Will we ever want to make it optional though?  I'd agree if so, but I never got that impression.

Reply:
It's totally possible, for example:
```java
double reduceSomething(Tensor input);
double reduceSomething(Tensor input, Tensor optionalOutput, Scope optionalScopeForOptionalOutput);
```
Do you see an advantage in having it as the first argument though?

Reply:
@karllessard @Craigacp Any preferences?

Reply:
I'm not sure if we need to enforce the explicit scoping of tensors resulting from an eager operation like this in all cases. Default behavior would tie these tensors to their eager session, which will release them once closed. Do we want to offer `TensorScope` as a simple utility or as something mandatory? I was under the impression of the former, meaning that we should leave an endpoint accepting no `TensorScope`.

Related to the order of the arguments in the methods, I don't have a very firm opinion on this but if I need to pick one, I also prefer having optional arguments at the end of the signature rather than the beginning.

Reply:
It would make most sense to me for an eager session to be a TensorScope, but we'd probably want to have it without attach() and detach(). Maybe we could have a special MutableTensorScope if we really want to have something with attach() and detach()?

Reply:
Agreed, but I'd just use composition instead of inheritance, there's no need to expose it.  Eventually I think it would make sense to tie the tensor lifetime to the lifetime of the eager operand, but since we don't have operand lifetimes yet this is fine.  I just don't like using EagerSession for scope management very much since it is usually very long lived, but if we have an optional tensor scope version it's fine for now.

For the first vs last argument, I was under the impression that the scope would be mandatory, like I thought Panama's was, but looking at it again I'm not actually sure.  If we keep detach (see other reply), I'm not sure what benefit optional scopes would really provide, other than removing some boilerplate.   If we do decide to use optional scopes, I agree about the last argument, I made it first since it is (was?) a required scope, and those live on the left, at least in Kotlin.  It really should be a receiver.

Reply:
OK let's try to see how it goes by enforcing a scope for the tensors, which is a better safety net than the GC for sure since these tensors can be quite large, compared to the operands native objects.